keps/sig-node/0028-20180906-quotas-for-ephemeral-storage.md

```diff
@@ -0,0 +1,647 @@
+  files being held open.
+* Primary: this will not interfere with the more common user and group
+  quotas.
+* Stretch: enforce limits on per-volume storage consumption by using
+  enforced project quotas.  Each volume would be given an enforced
+  quota of the total ephemeral storage limit of the pod.
+
+### Non-Goals
+
+* Enforcing limits on total pod storage consumption by any means, such
+  that the pod would be hard restricted to the desired storage limit.
+
+## Proposal
+
+This proposal applies project quotas to emptydir volumes on qualifying
+filesystems (ext4fs and xfs with project quotas enabled).  Project
+quotas are applied by selecting an unused project ID (a 32-bit
+unsigned integer), setting a limit on space and/or inode consumption,
+and attaching the ID to one or more files.  By default (and as
+utilized herein), if a project ID is attached to a directory, it is
+inherited by any files created under that directory.
+
+If we elect to use the quota as enforcing, we impose a quota
+consistent with the desired limit.  If we elect to use it as
+non-enforcing, we impose a large quota that in practice cannot be
+exceeded (2^63-1 bytes for XFS, 2^58-1 bytes for ext4fs).
+
+### Operation Flow -- Applying a Quota
+
+* Caller (emptydir volume manager or container runtime) creates an
+  emptydir volume, with an empty directory at a location of its
+  choice.
+* Caller requests that a quota be applied to a directory.
+* Determine whether a quota can be imposed on the directory, by asking
+  each quota provider (one per filesystem type) whether it can apply a
+  quota to the directory.  If no provider claims the directory, an
+  error status is returned to the caller.
+* Select an unused project ID (see [below](#selecting-a-project-id)).
+* Set the desired limit on the project ID, in a filesystem-dependent
+  manner (see [below](#notes-on-implementation)).
+* Apply the project ID to the directory in question, in a
+  filesystem-dependent manner.
+
+An error at any point results in no quota being applied and no change
+to the state of the system.  The caller in general should not assume a
+priori that the attempt will be successful.  It could choose to reject
+a request if a quota cannot be applied, but at this time it will
+simply ignore the error and proceed as today.
+
+### Operation Flow -- Retrieving Storage Consumption
+
+* Caller (kubelet metrics code, cadvisor, container runtime) asks the
+  quota code to compute the amount of storage used under the
+  directory.
+* Determine whether a quota applies to the directory, in a
+  filesystem-dependent manner (see [below](#notes-on-implementation)).
+* If so, determine how much storage or how many inodes are utilized,
+  in a filesystem dependent manner.
+
+If the quota code is unable to retrieve the consumption, it returns an
+error status and it is up to the caller to utilize a fallback
+mechanism (such as the directory walk performed today).
+
+### Operation Flow -- Removing a Quota.
```

Comment:
When do we remove a quota?

Reply:
When a volume is torn down (including at kubelet restart).